pkg/kube/dialer.go

```diff
@@ -0,0 +1,336 @@
+package kube
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/httpstream"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/transport/spdy"
+)
+
+const (
+	defaultClusterDomain = "cluster.local"
+	defaultNamespace     = "default"
+	svcLabel             = "svc"
+
+	portForwardProtocolV1Name = "portforward.k8s.io"
+)
+
+var (
+	ErrInvalidPodFQDN = errors.New("invalid pod FQDN")
+	ErrNoPort         = errors.New("no port specified")
+
+	negotiatedSerializer = serializer.NewCodecFactory(runtime.NewScheme()).WithoutConversion()
+)
+
+// PodDialer is a basic port-forwarding dialer that doesn't start
+// any local listeners, but returns a net.Conn directly.
+type PodDialer struct {
+	config        *rest.Config
+	clusterDomain string
+	requestID     int
+
+	connections map[types.NamespacedName]httpstream.Connection
+	mutex       sync.RWMutex
+}
+
+// NewPodDialer create a PodDialer.
+func NewPodDialer(config *rest.Config) *PodDialer {
+	return &PodDialer{
+		config:        config,
+		clusterDomain: defaultClusterDomain,
+		connections:   make(map[types.NamespacedName]httpstream.Connection),
+	}
+}
+
+// WithClusterDomain overrides the domain used for FQDN parsing.
+func (p *PodDialer) WithClusterDomain(domain string) *PodDialer {
+	p.clusterDomain = domain
+	return p
+}
+
+// Dial dials the given pod's service-based DNS address and returns a
+// net.Conn that can be used to reach the pod directly.
+func (p *PodDialer) Dial(network string, address string) (net.Conn, error) {
+	switch network {
+	case "tcp", "tcp4", "tcp6":
+	default:
+		return nil, fmt.Errorf("dialer only supports TCP-based networks: %w", net.UnknownNetworkError(network))
+	}
+
+	pod, port, err := p.parseDNS(address)
+	if err != nil {
+		return nil, err
+	}
+
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+
+	conn, err := p.connectionForPod(pod)
+	if err != nil {
+		return nil, err
+	}
+
+	p.requestID++
+
+	headers := http.Header{}
+	headers.Set(corev1.PortHeader, strconv.Itoa(port))
+	headers.Set(corev1.PortForwardRequestIDHeader, strconv.Itoa(p.requestID))
+
+	headers.Set(corev1.StreamType, corev1.StreamTypeError)
+	errorStream, err := conn.CreateStream(headers)
+	if err != nil {
+		return nil, err
+	}
+
+	headers.Set(corev1.StreamType, corev1.StreamTypeData)
+	dataStream, err := conn.CreateStream(headers)
+	if err != nil {
+		return nil, err
+	}
+
+	return wrapConn(network, address, dataStream, errorStream), nil
+}
+
+// DialContext is a simple wrapper around Dial.
+func (p *PodDialer) DialContext(_ context.Context, network string, address string) (net.Conn, error) {
```

Comment:
Explicitly comment that `context.Context` is NOT respected here. There doesn't seem to be a super straightforward way to plumb context aware into spdy :/ that's going to come back and bite us but better to have it in code we own instead of hiding in between k8s code and an unrelated proxy connection.

Reply:
I think I thought of a way where we can actually respect the Context by using a polling goroutine in the underlying connection implementation to just forcibly (and idempotently) Close the underlying stream. I'll push up the changes in a bit as I work through the rest of the comments here.